### ISSUE TITLE
IllegalType extends check is too strict

### ISSUE THREAD

**Issue by huxi** [checkstyle/checkstyle#6267]
#6118 introduced checking for illegal types in extends/implements.

The problem is that code like the following is a common way to work around Java generics type erasure:

```java
public class StringFooMap
	extends HashMap<String, Foo> {}
```

See [this Jackson issue](https://github.com/FasterXML/jackson-databind/issues/1606#issuecomment-294570555) for an example.

It would be nice if you could add an option to disable checking of extends.

````
$ cat TestClass.java
public class StringFooMap
    extends HashMap<String, Foo> {}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="IllegalType">
</module>
    </module>
</module>

$ java -jar checkstyle-8.15-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:2:13: Usage of type 'HashMap' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.
````

**Comment by rnveach** [comment]
@huxi This check does support XPath suppression, so you can always use that to suppress any violation you don't agree with.
See http://checkstyle.sourceforge.net/config_filters.html#SuppressionXpathFilter

You can also remove `CLASS_DEF` as a token, but this would also hide possible violations on direct type parameters and implements.

**Comment by romani** [comment]
interesting case, it is also reported at https://github.com/checkstyle/checkstyle/pull/6266

I think it does not matter where you use Illegal type in this case
```
public class StringFooMap
	extends HashMap<String, Foo> {}
```
vs
```
public class StringFooHashMap
	extends MyMap<String, HashMap> {}
```

I do not think there is way out to satisfy all. Term "illegal" is subjective and here are corner cases.

@huxi , is removing of CLASS_DEF token from Check `tokens` property a solution/workaround for you ? 
What about Xpath ? Do you need help to setup xpath suppression ?

**Comment by huxi** [comment]
@romani Quick feedback: I'm currently on vacation so I can't really work on this until mid-january. Sorry about that.

XPath suppression won't be a solution since there have been numerous errors like this in our code base because of heavy jackson use. I have no idea how to remove the CLASS_DEF token from the tokens property of the check but will find out and give that a try when I'm back at work.

In the meantime, I wish you a merry Christmas and a happy new year. Thanks for putting your time into checkstyle.

**Comment by romani** [comment]
thanks a lot for notification.

> I have no idea how to remove the CLASS_DEF token from the tokens property of the check but will find out and give that a try when I'm back at work.

If you are opensource project - give us a link to checkstyle config , we will help.

**Comment by huxi** [comment]
Closed source, unfortunately. But thanks for the offer.

I went ahead and suppressed the warning for the files in question.

**Comment by romani** [comment]
we should share a config of SimpleXpathSuppressionFilter (ones we finish it) and offer people single way suppress such cases.

**Comment by rnveach** [comment]
@huxi 

> XPath suppression won't be a solution since there have been numerous errors like this in our code base because of heavy jackson use.

Maybe I'm mis-understanding, but xpath isn't just limited to 1 file. You can trigger it on a specific extended class for a specific class.

I have provided an example using `SuppressionXpathSingleFilter` which isn't released, but the same fields are available using `SuppressionXpathFilter` ( http://checkstyle.sourceforge.net/config_filters.html#SuppressionXpathFilter ).  You can use the command line option `--generate-xpath-suppression` to help with generating xpaths from violations. ( http://checkstyle.sourceforge.net/cmdline.html#Command_line_usage )

````
$ cat TestClass.java
public class StringFooMap
extends HashMap<String, Foo> {}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="IllegalType">
</module>
<module name="SuppressionXpathSingleFilter">
  <property name="checks" value="IllegalType"/>
  <property name="query" value="//EXTENDS_CLAUSE[@text='HashMap']/IDENT"/>
</module>
    </module>
</module>

$ java -jar checkstyle-pr-suppression_xpath_single_filter-2019-01-06-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

> I have no idea how to remove the CLASS_DEF token from the tokens property of the check

Please see the following example:

````
$ cat TestClass.java
public class StringFooMap
extends HashMap<String, Foo> {}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="IllegalType">
  <property name="tokens" value="ANNOTATION_FIELD_DEF, INTERFACE_DEF, METHOD_CALL, METHOD_DEF, METHOD_REF, PARAMETER_DEF, VARIABLE_DEF" />
</module>
    </module>
</module>

$ java -jar checkstyle-8.17-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

**Comment by huxi** [comment]
@rnveach Ah, I see. I thought there was a way to really *remove* `CLASS_DEF` instead of re-listing all the default tokens beside the unwanted one. I refrained from doing it that way because I was worried that this default list may change in a future version of checkstyle without me noticing it.

I convinced myself in the meantime that I'd rather explicitly add files like the `StringFooMap` above to our `suppressions.xml` file than risking that ordinary code extends types like `HashMap` arbitrarily.

Thanks for the clarification.

**Comment by romani** [comment]
@huxi , pease close the issue, looks like suppression in any form is the only way to resolve such violations.
